Mathlib/Tactic/StacksAttribute.lean

```diff
@@ -0,0 +1,113 @@
+
+/-!
+# The `stacks` attribute
+
+This allows tagging of mathlib lemmas with the corresponding
+[Tags](https://stacks.math.columbia.edu/tags) from the Stacks Project.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Stacks
+
+/-- `Tag` is the structure that carries the data of a Stacks Projects tag and a corresponding
+Mathlib declaration. -/
+structure Tag where
+  /-- The Stacks Project tag. -/
+  tag : String
+  /-- The name of the declaration with the given tag. -/
+  declName : Name
+  deriving BEq, Hashable
+
+/-- Defines the `tagExt` extension for adding a `HashSet` of `Tag`s
+to the environment. -/
+initialize tagExt : SimplePersistentEnvExtension Tag (HashSet Tag) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := fun as => as.foldl HashSet.insertMany {}
+    addEntryFn := .insert
+  }
+
+/--
+`addTagEntry tag declName` takes as input the `String` `tag` and the `Name` `declName` of
+a declaration with the `stacks` attribute.
+It extends the `Tag` environment extension with the data `tag, declName`.
+-/
+def addTagEntry {m : Type → Type} [MonadEnv m] (tag : String) (declName : Name) : m Unit :=
+  modifyEnv (tagExt.addEntry · { tag := tag, declName := declName })
+
+/--
+The syntax for a Stacks tag: it is an optional number followed by an optional identifier.
+This allows `044Q3` and `GH3F6` as possibilities.
+-/
+declare_syntax_cat stackTag
+
+@[inherit_doc Parser.Category.stackTag]
+syntax (num)? (ident)? : stackTag
+
+/-- The `stacks` attribute.
+Use it as `@[stacks TAG "Optional comment"]`.
+The `TAG` is mandatory.
+
+See the [Tags page](https://stacks.math.columbia.edu/tags) in the Stacks project for more details.
+-/
+syntax (name := stacks) "stacks " (stackTag)? (ppSpace str)? : attr
+
+initialize Lean.registerBuiltinAttribute {
+  name := `stacks
+  descr := "Apply a Stacks project tag to a theorem."
+  add := fun decl stx _attrKind => Lean.withRef stx do
+    -- check that the tag consists of 4 characters and
+    -- that only digits and uppercase letter are present
+    let tag := stx[1]
+    match tag.getSubstring? with
+      | none => logWarning "Please, enter a Tag after `stacks`."
+      | some str =>
+        let str := str.toString.trimRight
+        if str.length != 4 then
+          logWarningAt tag
+            m!"Tag '{str}' is {str.length} characters long, but it should be 4 characters long"
+        else if 2 ≤ (str.split (fun c => (!c.isUpper) && !c.isDigit)).length then
+          logWarningAt tag m!"Tag '{str}' should only consist of digits and uppercase letters"
+        else match stx with
+          | `(attr| stacks $_:stackTag $_:str) => addTagEntry str decl
+          | `(attr| stacks $_:stackTag) => addTagEntry str decl
+          | _ => throwUnsupportedSyntax
+}
+
+end Mathlib.Stacks
+
+open Mathlib.Stacks
+/--
+`getSortedTags env` returns the array of `Tags`, sorted by alphabetical order of declaration name.
+-/
+def Lean.Environment.getSortedTags (env : Environment) : Array Tag :=
```

Comment:
This is big ambiguous from the name. I think most people would guess it returns the tags sorted by the stack project tag, like the master list on the stacks project repo. 

I'm guessing there will be more querying using the stacks project tag than the name of the registered declaration so adding a function here to do that would nice also.

Reply:
Ok, I'll rename to `Lean.Environment.getSortedStackProjectTags`.

The second comment is about creating a function that takes a (string) "tag" as input and returns an `Option Name`?

Reply:
I think you should just switch to returning the `Stacks.Tag`s sorted by the `tag` rather than `declName`. 

The second comment was about a `Lean.Environment.findStacksTag` function which indeed returned an `Option Name` and consumed the string Stacks project tag.

Reply:
Ok, great!  I'll return an `Array Name`, since conceivably the same Stacks tag could be applied to multiple lemmas, if they each prove a part of the Stacks retult.

Reply:
I just pushed both changes